homeassistant/components/zwave_js/migrate.py

```diff
@@ -1,27 +1,354 @@
+        """Return data to save."""
+        return self._data
+
+    @callback
+    def add_entity_value(
+        self,
+        config_entry: ConfigEntry,
+        entity_id: str,
+        discovery_info: ZwaveDiscoveryInfo,
+    ) -> None:
+        """Add info for one entity and Z-Wave JS value."""
+        ent_reg = async_get_entity_registry(self._hass)
+        dev_reg = async_get_device_registry(self._hass)
+
+        node = discovery_info.node
+        primary_value = discovery_info.primary_value
+        entity_entry = ent_reg.async_get(entity_id)
+        assert entity_entry
+        device_identifier = get_device_id(node.client, node)
+        device_entry = dev_reg.async_get_device({device_identifier}, set())
+        assert device_entry
+
+        # Normalize unit of measurement.
+        unit = entity_entry.unit_of_measurement
+        unit = unit and unit.lower()
+        if unit == "":
+            unit = None
+
+        data: ZWaveJSMigrationData = {
+            "node_id": node.node_id,
+            "endpoint_index": node.index,
+            "command_class": primary_value.command_class,
+            "value_property_name": primary_value.property_name,
+            "value_property_key_name": primary_value.property_key_name,
+            "value_id": primary_value.value_id,
+            "device_id": device_entry.id,
+            "domain": entity_entry.domain,
+            "entity_id": entity_id,
+            "unique_id": entity_entry.unique_id,
+            "unit_of_measurement": unit,
+        }
+
+        self.save_data(config_entry.entry_id, entity_id, data)
+
+    async def get_data(
+        self, config_entry: ConfigEntry
+    ) -> dict[str, ZWaveJSMigrationData]:
+        """Return Z-Wave JS migration data for a config entry."""
+        await self.load_data()
+        data = self._data.get(config_entry.entry_id)
+        return data or {}
+
+
+@callback
+def async_map_legacy_zwave_values(
+    zwave_data: dict[str, ZWaveMigrationData],
+    zwave_js_data: dict[str, ZWaveJSMigrationData],
+) -> LegacyZWaveMappedData:
+    """Map Z-Wave node values onto Z-Wave JS node values."""
+    migration_map = LegacyZWaveMappedData()
+    zwave_proc_data: dict[
+        tuple[int, int, int, str, str | None, str | None],
+        ZWaveMigrationData,
+    ] = {}
+    zwave_js_proc_data: dict[
+        tuple[int, int, int, str, str | None, str | None],
+        ZWaveJSMigrationData,
+    ] = {}
+
+    for zwave_item in zwave_data.values():
+        zwave_js_property_name = CC_ID_LABEL_TO_PROPERTY.get(
+            zwave_item["command_class"], {}
+        ).get(zwave_item["command_class_label"])
+        item_id = (
+            zwave_item["node_id"],
+            zwave_item["command_class"],
+            zwave_item["node_instance"] - 1,
+            zwave_item["domain"],
+            zwave_item["unit_of_measurement"],
+            zwave_js_property_name,
+        )
+
+        # Remove duplicates that are not resolvable.
```

Comment:
Would it be possible for triple duplicates to exist? Because a third occurrence will now be added because we removed the key from `zwave_proc_data`. Alternative is to set it to `None` and filter it out later.

Reply:
Good point. I'll change it.